r.curvenumber: add antecedent runoff condition and hydrologic condition options

mabdazzam opened the pull request:
1. Additional options of two conditions, Hydrologic Condition (Poor, Fair, Good) and Antecedent Runoff Condition (I, II, III).
2. Updated lookup tables for ESA and NLCD to incorporate HCs and ARCs and documentation.

mabdazzam committed:
add hc and arc conditions - update doc

mabdazzam committed:
Ruff formatting fixes

mabdazzam committed:
fix ruff indentation and linting issue in markown

mabdazzam committed:
lin fix; top level heading in the beginning, 80 characters limit, trailing space

petrasovaa commented on src/raster/r.curvenumber/r.curvenumber.md:
```diff
@@ -1,8 +1,12 @@
+# r.curvenumber
+
```
```suggestion
```

petrasovaa commented on src/raster/r.curvenumber/r.curvenumber.md:
````diff
@@ -36,28 +51,32 @@ assumptions, and spatial scale of their hydrologic studies or engineering design
 ```
 
-![r.curvenumber example](r_curvenumber_output.png)  
+![r.curvenumber example](r_curvenumber_output.png)
````
This is newline, should stay there:
```suggestion
![r.curvenumber example](r_curvenumber_output.png)  
```

petrasovaa commented on src/raster/r.curvenumber/r.curvenumber.py:
```diff
@@ -42,7 +42,25 @@
+# % key: hydrologic_condition
+# % type: string
+# % description: Hydrologic condition
+# % options: poor,fair,good
```
Please add descriptions, what is poor, fair, good hydrologic condition? Would this be correct?

```suggestion
# % options: poor,fair,good
# % descriptions: poor;increased runoff;fair;normal runoff;good;lower runoff
```

mabdazzam replied:
Poor: <30% of groundcover (bare soil), Fair: 30% - 70% groundcover (partial cover), Good: >70% groundcover (dense cover), which translates to what you described in terms of runoff so I think it is correct.

petrasovaa commented on src/raster/r.curvenumber/r.curvenumber.py:
```diff
@@ -42,7 +42,25 @@
+# % key: antecedent_runoff_condition
+# % type: string
+# % description: Antecedent Runoff Condition (ARC)
+# % options: i,ii,iii
```
Consider using dry/normal/wet instead or at least explain it here:

```suggestion
# % description: Antecedent Runoff Condition (the degree of wetness of a watershed before a rainfall)
# % options: i,ii,iii
# % descriptions: i;dry;ii;normal;iii;wet
```

petrasovaa commented on src/raster/r.curvenumber/r.curvenumber.py:
```diff
@@ -51,159 +69,1121 @@
+41,2,fair,i,39
+41,3,fair,i,53
+41,4,fair,i,60
+41,1,good,i,12
```
ChatGPT complains 12 is too low, should be at least 30. Any opinion?

mabdazzam replied:
So there are two options to handle this. The minimum curve number recommended (by the SCS Method) for runoff computation is 30, but the standard tables do include CN < 30 for some land uses.  
Should I enforce this floor in the runoff module itself, or handle it here by setting any CN in (0, 30) to 30?

mabdazzam replied:
I have updated the minimum CN to 30, since CN is used only for runoff computation, there's no point in providing the values which are not valid for runoff estimation.

petrasovaa commented on src/raster/r.curvenumber/r.curvenumber.py:
```diff
@@ -51,159 +69,1121 @@
-100,3,78
-100,4,79
+ESA_CSV = """lc,hsg,hc,arc,cn
+10,1,poor,i,45
+10,2,poor,i,66
+10,3,poor,i,77
+10,4,poor,i,83
+10,1,poor,ii,26
+10,2,poor,ii,46
+10,3,poor,ii,59
+10,4,poor,ii,67
+10,1,poor,iii,65
+10,2,poor,iii,82
+10,3,poor,iii,89
+10,4,poor,iii,93
```
The value for dry conditions is higher than for fair conditions, seems wrong, no? This is not the only place with this issue.

```
10,1,poor,i,45
10,1,poor,ii,26
10,1,poor,iii,65

mabdazzam replied:
looks like I misread the original ESA table and swapped the ARC I and ARC II values. I’ll fix it.

petrasovaa commented on src/raster/r.curvenumber/r.curvenumber.py:
```diff
@@ -51,159 +69,1121 @@
+                    row["hc"].lower(),
+                    row["arc"].lower(),
+                ] = row["cn"]
+            # lut[row["lc"], row["hsg"]] = row["cn"]
```
```suggestion
```

petrasovaa commented on src/raster/r.curvenumber/r.curvenumber.py:
```diff
@@ -51,159 +69,1121 @@
+        lut[
+            row["lc"].lower(), row["hsg"].lower(), row["hc"].lower(), row["arc"].lower()
+        ] = row["cn"]
+        # lut[row["lc"], row["hsg"]] = row["cn"]
```
```suggestion
```

petrasovaa commented on src/raster/r.curvenumber/r.curvenumber.py:
```diff
@@ -212,33 +1192,45 @@ def load_custom(path):
         lut = load_custom(custom)
     else:
-        fatal("Must specify --source=nlcd|esa or provide --lookup for custom")
+        fatal(_("Must specify --source=nlcd|esa or provide --lookup for custom"))
```
```suggestion
        fatal(_("Must specify --source=nlcd|esa or provide lookup for custom"))
```

petrasovaa reviewed (changes requested):
How did you actually get all these numbers? I don't see the ARC anywhere in the published tables, could we just adjust the number dynamically based on the selected ARC?

mabdazzam committed:
Update src/raster/r.curvenumber/r.curvenumber.md

I added this because Linting and Formatting test was failing without a top level heading in the beginning. 
It appears to be caused by the newer Ruff (v0.12.0).

Co-authored-by: Anna Petrasova <kratochanna@gmail.com>

mabdazzam committed:
Update src/raster/r.curvenumber/r.curvenumber.md

Co-authored-by: Anna Petrasova <kratochanna@gmail.com>

mabdazzam committed:
Update src/raster/r.curvenumber/r.curvenumber.py

Co-authored-by: Anna Petrasova <kratochanna@gmail.com>

mabdazzam commented:
> How did you actually get all these numbers? I don't see the ARC anywhere in the published tables, could we just adjust the number dynamically based on the selected ARC?

Actually, the published Curve Number tables in Chapter 9 only list values for ARC II. To get the corresponding values for ARC I and ARC III, you use the adjustment factors in Table 10-1 (“Estimation of Direct Runoff from Storm Rainfall,” 210-VI-NEH, Chapter 10). You can view Table 10-1 here:
https://directives.sc.egov.usda.gov/landingpage/dea0a36a-e5ba-4de2-83d7-d7b6a8f9f9e6

I’ve also updated the markdown reference to include Chapter 10.

mabdazzam committed:
accepted minor fixes, fixed the swapped arc i values with arc ii, added explanation of ARCs

mabdazzam committed:
add reference to ARC II to ARC I and III conversion table from handbook chapter 10

petrasovaa commented:
> > How did you actually get all these numbers? I don't see the ARC anywhere in the published tables, could we just adjust the number dynamically based on the selected ARC?
> 
> Actually, the published Curve Number tables in Chapter 9 only list values for ARC II. To get the corresponding values for ARC I and ARC III, you use the adjustment factors in Table 10-1 (“Estimation of Direct Runoff from Storm Rainfall,” 210-VI-NEH, Chapter 10). You can view Table 10-1 here: https://directives.sc.egov.usda.gov/landingpage/dea0a36a-e5ba-4de2-83d7-d7b6a8f9f9e6
> 
> I’ve also updated the markdown reference to include Chapter 10.

Thanks, that's helpful. I think we should have that table in the code explicitly and separately from the other tables. I know it's more work, but it will make the other tables shorter and more maintainable.

mabdazzam committed:
add separate table and functionality for arc conversion

mabdazzam committed:
add explanation of hsg, hc, arc nad update references in the documentation

mabdazzam committed:
fix ruff formatting issues

mabdazzam committed:
fix ruff formatting issues

mabdazzam committed:
changes by pre-commit suite

petrasovaa commented on src/raster/r.curvenumber/r.curvenumber.py:
```diff
@@ -42,7 +42,27 @@
+# %option
+# % key: hydrologic_condition
+# % type: string
+# % description: Hydrologic condition (groundcover density afftecting runoff potential)
```
```suggestion
# % description: Hydrologic condition (groundcover density affecting runoff potential)
```

mabdazzam replied:
fixed

petrasovaa commented on src/raster/r.curvenumber/r.curvenumber.py:
```diff
@@ -51,194 +71,548 @@
+60,1,good,65
+60,2,good,79
+60,3,good,87
+60,4,good,90
```
Lot of these values are the same for poor/fair/good. Is that intention?

mabdazzam replied:
Yes, it's intentional. This class represents bare soil, so hydrologic condition doesn’t apply; there’s no vegetative cover to vary by percentage. The same CN is used across poor/fair/good to reflect that.

petrasovaa commented on src/raster/r.curvenumber/r.curvenumber.py:
```diff
@@ -51,194 +71,548 @@
+                    "in fair condition are interpolated as"
+                    "the average of poor and good conditions"
+                )
+            )
```
This could be simplified, no? You don't need to test for the specific lc, we know it's going to be always there. You could check whether that value is in the input raster, but that would look differently than this. I am more leaning towards removing this and adding a note in the documentation.

mabdazzam replied:
Agreed. Moved it to documentation.

petrasovaa commented on src/raster/r.curvenumber/r.curvenumber.py:
```diff
@@ -51,194 +71,548 @@
+            expr = f"if({landmap}=={lc} && {hsg}=={grp}, {adjusted_cn}, {expr})"
+
+    # enforce minimum CN of 30
+    expr = f"if({expr} < 30, 30, {expr})"
```
```suggestion
    expr = f"max({expr}, 30)"
```

mabdazzam replied:
I have updated this check in a different way, I want to set 0 < CN < 30 to 30. Sometimes zero is a valid CN (say ice).

petrasovaa reviewed (changes requested):
Almost there!

mabdazzam committed:
Update src/raster/r.curvenumber/r.curvenumber.py

Co-authored-by: Anna Petrasova <kratochanna@gmail.com>

mabdazzam committed:
remove lc=40 warning, update minimum useful CN to 30, fix typo

mabdazzam committed:
add lc 40 note in the documentation

mabdazzam committed:
fix linting issue; add blank line before the heading

petrasovaa reviewed (approved):
I think this is good to go! Thank you
